### PR TITLE
SMART connect "fail" callback and API Verifier standalone

### PR DESCRIPTION
**Pull request by nschwertner**

### CODE REVIEW

File: static/framework/smart/scripts/smart-api-client-base.js [hunk 1]
```diff
@@ -12,14 +12,39 @@ var SMART_CONNECT_CLIENT = function(smart_server_origin, frame) {
     var channel = null;
 
     this.is_ready = false;
```
Comment (jmandel) [inline static/framework/smart/scripts/smart-api-client-base.js]: I know this is my doing -- but these should be private state, not public on the SMART object (i.e. we shouldn't expose`SMART.is_ready`)

File: static/framework/smart/scripts/smart-api-client-base.js [hunk 2]
```diff
@@ -12,14 +12,39 @@ var SMART_CONNECT_CLIENT = function(smart_server_origin, frame) {
...
+            this.fail_callback();
+        }
+        return this;
+    };
+    
+    this.timeoutProtect = setTimeout(function() {
+        _this.timeoutProtect = null;
+        _this.has_failed = true;
+        if (_this.fail_callback) _this.fail_callback();
+    }, 5000);
```
Comment (jmandel) [inline static/framework/smart/scripts/smart-api-client-base.js]: This timer kicks off when `SMART.fail()` is first called.  My understanding is the channel's construction won't be blocked by slow-loading page resources (that is, nothing is tied to an `onload` event)?  If so 5s seems fine (generous).

**Comment by jmandel** [comment]
Graceful degradation of the API Verifier (limited tabs) looks awesome!

Re: `smart-api-client.js` I have just a few questions/clarifications (above). 

Also worth deciding: should we be using the jQuery deferred syntax here, too, as in:

```
SMART.ready
     .success(function(){
        console.log("Hurray!");
     })
    .error(function(){
        console.log("Sadness.");
     })
```

Thoughts?

File: static/framework/smart/scripts/smart-api-client-base.js (hunk 2 above)
Comment (nschwertner) [inline static/framework/smart/scripts/smart-api-client-base.js]: The timer is winded when you first load the SMART connect lib (just before launching the channel's constructor). You can call `SMART.fail()` at any time independently of the timer. Nothing is tied to `onload`. So, overall I think that 2-5s is about the right range for this timeout.

File: static/framework/smart/scripts/smart-api-client-base.js (hunk 1 above)
Comment (nschwertner) [inline static/framework/smart/scripts/smart-api-client-base.js]: I can fix this.

**Comment by nschwertner** [comment]
Hmm... an interesting question of form and consistency. My gut feeling is that `SMART.ready.success(...)` is a bit verbose and can confuse people. Also, I think that keeping the names of these callback handlers different from the ones we use the the API convenience methods highlights the differentiating nuance between them. In this case, the `ready` and `fail` methods are hooks for discovering the outcome of an event that you don't trigger explicitly (establishing a channel to the container) while the `success` and `error` methods are tied to an explicit call execution.

So, I am happy to keep the names as is. Then again, I can see the argument going the other way too.

**Comment by jmandel** [comment]
> So, I am happy to keep the names as is.

 I think this is fair (and also has the tie-breaking property of not
breaking existing code).

  -J

**Comment by jmandel** [comment]
This pull request was based on `master` -- we should be merging these topic branches into `dev`.  Reverted.

**Comment by nschwertner** [comment]
Yes, I noticed that for some reason GitHub sets the pull requests against master even when they have been branched off dev. For the more recent pull requests I had to manually update the target to "dev" when setting them up.

I will set up a new pull request for this code...

**Comment by jmandel** [comment]
No need -- it's already pulled along with #3 :-)

On Wed, May 16, 2012 at 9:48 AM, nschwertner
reply@reply.github.com
wrote:

**Pull request merged**